Add stub Gateway & GatewayClass controllers

david-martin opened the pull request:
Closes #42 

Stub controllers for division of upcoming work on Gateway & GatewayClass functionality.
No new functionality in this PR.

the `test` target now runs any controller integration tests (There's currently a few stubs for controllers).

Steps used to generate a controller and move them to the right location:

```
kubebuilder create api --group gateway.networking.k8s.io --version v1beta1 --kind Gateway
mv controllers pkg/controllers/gateway
```
The kubebuilder cmd failed with this error:
```
github.com/Kuadrant/multi-cluster-traffic-controller/controllers tested by
        github.com/Kuadrant/multi-cluster-traffic-controller/controllers.test imports
        sigs.k8s.io/controller-runtime/pkg/envtest/printer: module sigs.k8s.io/controller-runtime@latest found (v0.14.2), but does not contain package sigs.k8s.io/controller-runtime/pkg/envtest/printer
```
So I manually modified the suite to remove the envtest/printer dependency since it wasn't being used in the generated file.

Package names were modified manually to *not* be `controllers`, and the import of the controllers in main.go was updated to point to the right pkg dir.

The gateway api was imported to the controllers and the CRD referenced in SetupWithManager.

david-martin commented:
Depends on #31 to work out of the box.
Otherwise, ovewriting the webhook-port is enough to verify this PR.

maleck13 commented on pkg/controllers/gatewayclass/gatewayclass_controller.go:
```diff
@@ -0,0 +1,62 @@
+limitations under the License.
+*/
+
+package gatewayclass
```
I think we could move this under the gateway package as discussed

maleck13 reviewed (approved):
once we move the gatewayclass into the gateway package I think we are good to merge

maleck13 commented:
@david-martin might be worth lookinga t https://github.com/Kuadrant/multi-cluster-traffic-controller/pull/37 as this looks to add Gateway API to control plane also

david-martin commented:
> once we move the gatewayclass into the gateway package I think we are good to merge

Moved now.
Also removed the generated suite_test.go files.
Will open a separate PR with a solution for running controller integration tests via envtest.

david-martin commented:
@maleck13 rebased.
The envtest piece was simpler than I initially thought, so this PR adds the suite_test.go stub for the gateway pkg.
Now, running `make test` will also include any controller integration tests as well.

mikenairn commented on config/rbac/role.yaml:
```diff
@@ -55,6 +55,58 @@ rules:
   - update
   - watch
+- apiGroups:
+  - gateway.networking.k8s.io.kuadrant.io
```
is this right?

david-martin replied:
Good catch.
kubebuilder.
I'll fixup in a PR

david-martin replied:
https://github.com/Kuadrant/multi-cluster-traffic-controller/pull/57
